components/org.wso2.carbon.extension.identity.sso.jwt/src/main/java/org/wso2/carbon/identity/sso/jwt/factory/JWTInboundIdentityResponseFactory.java

```diff
@@ -0,0 +1,164 @@
+     *
+     * @param identityResponse incoming IdentityResponse from the identity framework
+     * @return true if the incoming response is of the type handled by this factory
+     */
+    @Override
+    public boolean canHandle(IdentityResponse identityResponse) {
+
+        return identityResponse instanceof JWTInboundResponse;
+    }
+
+    /**
+     * Converts the received IdentityResponse instance to an HTTPResponse so that it could be sent to the calling party.
+     * This is where the logic for picking up and setting any parameters/headers/cookies etc is written.
+     *
+     * @param identityResponse the received (and handle-able IdentityResponse instance
+     * @return a corresponding HTTPResponse in the form of a builder, so that it could be built on demand
+     */
+    @Override
+    public HttpIdentityResponse.HttpIdentityResponseBuilder create(IdentityResponse identityResponse) {
+
+        HttpIdentityResponse.HttpIdentityResponseBuilder builder
+                = new HttpIdentityResponse.HttpIdentityResponseBuilder();
+
+        if (identityResponse instanceof JWTInboundResponse) {
+            JWTInboundResponse jwtInboundResponse = (JWTInboundResponse) identityResponse;
+
+            String logoutUrl = jwtInboundResponse.getLogoutUrl();
+            if (StringUtils.isBlank(jwtInboundResponse.getToken()) && StringUtils.isNotBlank(logoutUrl)) {
+                // Successful logout scenario.
+                if (log.isDebugEnabled()) {
+                    log.debug("Logout URL: " + JWTInboundUtil.neutralize(logoutUrl) + " provided and the token is set" +
+                            " to empty. Hence considering as successful logout scenario.");
+                }
+                builder.setStatusCode(HttpServletResponse.SC_FOUND);
+                // Redirect to Logout URL.
+                builder.setRedirectURL(logoutUrl);
+            } else if (StringUtils.isBlank(jwtInboundResponse.getToken())) {
+                // un-authenticated scenario without logout URL - Redirect to Retry.do - Show Client Error
+                String clientErrorPage = jwtInboundResponse.getEndpointUrl();
+                if (StringUtils.isNotBlank(clientErrorPage)) {
+                    builder.setStatusCode(HttpServletResponse.SC_FOUND);
+                    builder.setRedirectURL(clientErrorPage);
+
+                    Map<String, String[]> parameters = new HashMap<>();
+                    if (jwtInboundResponse.getParameters() != null) {
+                        for (Map.Entry<String, String> entry : jwtInboundResponse.getParameters().entrySet()) {
+                            parameters.put(entry.getKey(), new String[]{entry.getValue()});
+                        }
+                        builder.setParameters(parameters);
+                    }
+                } else {
+                    builder.setStatusCode(HttpServletResponse.SC_UNAUTHORIZED);
```

Comment:
Is this actually a 401 unauthorised error or is it added here as a generic error?

Reply:
Changed to 400